edn/src/matcher.rs

```diff
@@ -0,0 +1,576 @@
+        assert_match!("{_ _, _ _, 5 6}" ~= "{3 4, 1 2, 5 6}", false);
+    }
+
+    #[test]
+    fn test_match_placeholder_simple() {
+        assert_match!("?x" ~= "nil", true);
+        assert_match!("?x" ~= "true", true);
+        assert_match!("?x" ~= "1", true);
+        assert_match!("?x" ~= "1N", true);
+        assert_match!("?x" ~= "1.0", true);
+        assert_match!("?x" ~= "\"a\"", true);
+        assert_match!("?x" ~= "_", true);
+        assert_match!("?x" ~= "symbol", true);
+        assert_match!("?x" ~= "ns/symbol", true);
+        assert_match!("?x" ~= ":keyword", true);
+        assert_match!("?x" ~= ":ns/keyword", true);
+        assert_match!("?x" ~= "[nil, true, 1, \"foo\", bar, :baz]", true);
+        assert_match!("?x" ~= "(nil, true, 1, \"foo\", bar, :baz)", true);
+        assert_match!("?x" ~= "#{nil, true, 1, \"foo\", bar, :baz}", true);
+        assert_match!("?x" ~= "{nil true, 1 \"foo\", bar :baz}", true);
+    }
+
+    #[test]
+    fn test_match_placeholder_in_same_collection_type_simple() {
+        assert_match!("[?x]" ~= "[1]", true);
+        assert_match!("(?x)" ~= "(2)", true);
+        assert_match!("#{?x}" ~= "#{3}", true);
+        assert_match!("{?x ?x}" ~= "{4 4}", true);
+        assert_match!("{?x ?x}" ~= "{4 5}", false);
+        assert_match!("{?x ?y}" ~= "{4 4}", true);
+        assert_match!("{?x ?y}" ~= "{4 5}", true);
+    }
+
+    #[test]
+    fn test_match_placeholder_in_different_collection_type_simple() {
+        assert_match!("[?x]" ~= "(1)", false);
+        assert_match!("(?x)" ~= "#{2}", false);
+        assert_match!("#{?x}" ~= "[3]", false);
+        assert_match!("{?x ?x}" ~= "[4 5]", false);
+        assert_match!("{?x ?x}" ~= "(6 7)", false);
+        assert_match!("{?x ?x}" ~= "#{8 9}", false);
+    }
+
+    #[test]
+    fn test_match_placeholder_in_vector_with_multiple_values() {
+        assert_match!("[?x ?y]" ~= "[1 2]", true);
+        assert_match!("[?x ?y]" ~= "[1 1]", true);
+        assert_match!("[?x ?x]" ~= "[1 2]", false);
+        assert_match!("[?x ?x]" ~= "[1 1]", true);
+
+        assert_match!("[1 ?x 3 ?y]" ~= "[1 2 3 4]", true);
+        assert_match!("[1 ?x 3 ?y]" ~= "[1 2 3 2]", true);
+        assert_match!("[1 ?x 3 ?x]" ~= "[1 2 3 4]", false);
+        assert_match!("[1 ?x 3 ?x]" ~= "[1 2 3 2]", true);
+
+        assert_match!("[1 [2 [3 ?x]] 5 [?y 7]]" ~= "[1 [2 [3 4]] 5 [6 7]]", true);
+        assert_match!("[1 [2 [3 ?x]] 5 [?y 7]]" ~= "[1 [2 [3 4]] 5 [4 7]]", true);
+        assert_match!("[1 [2 [3 ?x]] 5 [?x 7]]" ~= "[1 [2 [3 4]] 5 [6 7]]", false);
+        assert_match!("[1 [2 [3 ?x]] 5 [?y 7]]" ~= "[1 [2 [3 4]] 5 [4 7]]", true);
+
+        assert_match!("[?x ?y ?x ?y]" ~= "[1 2 1 2]", true);
+        assert_match!("[?x ?y ?x ?y]" ~= "[1 2 2 1]", false);
+
+        assert_match!("[[?x ?y] [?x ?y]]" ~= "[[1 2] [1 2]]", true);
+        assert_match!("[[?x ?y] [?x ?y]]" ~= "[[1 2] [2 1]]", false);
```

Comment:
I think that actually works with the current implementation, but I'll add a test to make sure.